### DISCUSSION TITLE
How to set boundary conditions for a newly created subdomain?

### DISCUSSION THREAD

**Discussion by PEI0214** [idaholab/moose#28549] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello
I want to compute only in the newly created subdomain. So I only set the variables in the newly created subfield. But when I set the boundary condition, I got the following error:
~~~
*** ERROR ***
'archor_x' of type 'FunctionDirichletBC' depends on variable(s) 'disp_y,disp_z'. However, that variable does not appear to be defined on (all of) boundary ''.
~~~
Is there any way to set boundary conditions for newly created subdomains?

**Comment by GiudGiud** [comment]
> for newly created subdomains?

do you mean newly created boundaries?
 What did you specify for the boundary of FunctionDirichletBC ?

**Comment by PEI0214** [comment]
For newly created subdomains refer to blocks.
~~~
add_subdomain_ids = '${new_block_ids}'
~~~
FunctionDirichletBC limits the displacement of the boundary.
The variables and boundary conditions are set as follows:
~~~
[Variables]
  [disp_x]
    block = '${new_block_ids}'
  []
  [disp_y]
    block = '${new_block_ids}'
  []
  [disp_z]
    block = '${new_block_ids}'
  []
[]

[BCs]

  [archor_x]
    type = FunctionDirichletBC
    boundary = 'bottom'
    variable = disp_x
    value = 0
  []

  [archor_y]
    type = FunctionDirichletBC
    boundary = 'bottom'
    variable = disp_y
    value = 0
  []

  [archor_z]
    type = FunctionDirichletBC
    boundary = 'bottom'
    variable = disp_z
    value = 0
  []
[]
~~~
The 'bottom' is the boundary of the initial grid.  Is there a way to add a boundary to a newly created block?

**Reply by GiudGiud** [reply]
You can create new boundary using this object
https://mooseframework.inl.gov/source/meshmodifiers/SidesetAroundSubdomainUpdater.html

and predeclare it using the `add_sideset_ids / names` parameter

**Reply by PEI0214** [reply]
That's not what I wanted. My grid and input files are as follows:
Moves the initial grid blocks to the newly created blocks. And set the boundary conditions. Set the displacement of the bottom boundary (including 'bottom, left1, left2, left3, right1, right2, right3') to 0.
Can you run it? And give some advice? Is there a way to set variables for boundaries?

**Reply by PEI0214** [reply]
@GiudGiud Hello，do you have any good suggestions?

**Reply by GiudGiud** [reply]
I ll take a look. I still dont understand what you really want but it might come

> Is there a way to set variables for boundaries?

Using dirichlet boundary conditions

**Comment by GiudGiud** [comment, marked as answer]
ok I think I know what you are trying to do.
There's an integrity check on the variable restrictions vs the dirichlet BCs boudnary restrictions
Set this parameter to bypass it
```
[Problem]
  boundary_restricted_node_integrity_check = false
[]
```

**Reply by PEI0214** [reply]
That solves the problem. Thank you very much for your help.

**Reply by PEI0214** [reply]
Hello
The following error occurred when I set the pressure. Do you have any good suggestions?
~~~
*** ERROR ***
Error constructing object 'wateface_pressure' while retrieving value for 'variable' parameter! Did you forget to set 'variable' or set it to '' (empty string) by accident?
~~~
The pressure boundary conditions are set as follows:
~~~
[Functions]
  [w_p1]
    type = ParsedFunction
    expression = 'max(1e3 * (t -10), 0)'
  []
[]

[BCs]
  [wateface_pressure]
    type = Pressure
    boundary = 'top'
    function = w_p1
  []
[]
~~~

**Reply by GiudGiud** [reply]
The error message is correct here I think. 
You are missing the variable parameter in this water_pressure BC

**Reply by PEI0214** [reply]
When I add a variable, I still get this error.

**Reply by GiudGiud** [reply]
the syntax for the Pressure action is wrong, it should not be nested under BCs, but under Pressure
see this page
https://mooseframework.inl.gov/syntax/BCs/Pressure/index.html

**Reply by PEI0214** [reply]
https://mooseframework.inl.gov/syntax/BCs/Pressure/index.html
The Pressure in this file should be under BCs. I still get an error when I add Pressure according to the syntax in the file.

**Reply by GiudGiud** [reply]
right but it should be
```
[BCs]
  [Pressure]
    [wateface_pressure]
      boundary = 'bottom'
      function = w_p1
      displacements = 'disp_x disp_y disp_z'
     []
  []
[]
```
based on the example

**Reply by PEI0214** [reply]
I did so. But that didn't work. Can you run it?

**Reply by GiudGiud** [reply]
what does your BC block look like now?

**Reply by PEI0214** [reply]
~~~
[BCs]
  [archor_x]
    type = DirichletBC
    boundary = 'bottom left1 left2 left3 right1 right2 right3'
    variable = disp_x
    value = 0
  []

  [archor_y]
    type = DirichletBC
    boundary = 'bottom left1 left2 left3 right1 right2 right3'
    variable = disp_y
    value = 0
  []

  [archor_z]
    type = DirichletBC
    boundary = 'bottom left1 left2 left3 right1 right2 right3'
    variable = disp_z
    value = 0
  []

  [Pressure]
    [pressure1]
      boundary = 'top'
      function = 'max(1000 * (t - 10) ,0)'
    []
  []
[]
~~~

**Reply by GiudGiud** [reply]
can you add the displacements ?
see my snippet

**Reply by PEI0214** [reply]
I set displacements at GlobalParams.
~~~
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]
~~~

**Reply by PEI0214** [reply]
Can you run it successfully by setting this boundary condition?
~~~
boundary_restricted_node_integrity_check = false
~~~
This doesn't seem to work for Pressure.

**Reply by GiudGiud** [reply]
pressure is not a nodal object, it derives from IntegratedBC
so it s not affected by this boolean

**Reply by PEI0214** [reply]
@GiudGiud Do you have any good suggestions to solve this problem? Is there a way to restrict IntegratedBC checks?

**Reply by GiudGiud** [reply]
not off the top of my head. I ll try a few things

**Reply by GiudGiud** [reply]
so you can pass
```
  boundary_restricted_elem_integrity_check = false
```
to get past the error, but then the BC will fail / error.

**Reply by GiudGiud** [reply]
you could try this
```
[Variables]
  [disp_x]
    # block = '${new_domain_blocks}'
  []
  [disp_y]
    # block = '${new_domain_blocks}'
  []
  [disp_z]
    # block = '${new_domain_blocks}'
  []
[]

[Kernels]
  [time-x]
    type = NullKernel
    variable = disp_x
  []
  [time-y]
    type = NullKernel
    variable = disp_y
  []
  [time-z]
    type = NullKernel
    variable = disp_z
  []
[]
```
that way the displacements are defined everywhere from the beginning. And you dont get 0 on the diagonal on the subdomains where they are not the solution of the solid mechanics equations

**Reply by GiudGiud** [reply]
if you are worried about the pressure BCs firing up too early, you can use the Controls system to turn it off then on at a given time / condition. It take a little bit of work to get the condition right
